src/pages/blog/dynamic-components-in-your-mpa.mdx

```diff
@@ -0,0 +1,248 @@
+---
+layout: ../../components/Layout/Layout.astro
+title: Using dynamic components in your MPA
+description: It's not just React!
+date: 1.30.24
+categories:
+  - Dynamic components
+  - MPA
+  - Architecture
+---
+import { Card } from "../../components/Card";
+
+You have a multi-page application (MPA) which renders your frontend and it meets 99% of your needs. Your HTML is rendered from a server-side templating language like ASP, Razor, JSP, Django Templates, Pug, Handlebars, etc. It's lightweight, fast, and ideal in its technical simplicity. Up until now, any dynamic behavior you've needed has been easy enough to add via progressive enhancement in vanilla JS using imperative programming. But then the newest business requirement comes in and it's a real doozy. You know immediately: this would be easy to do as part of an SPA using a frontend framework like React or Vue, but this feels like a huge ask under the current architecture.
+
+So you need to introduce some dynamic components to your MPA code-base. What should you do? This article aims to discuss different scenarios and how they can be handled in the order of most ideal to least ideal.
+
+## Scenario 1: Your existing stack already accounts for this!
+
+The first step I can recommend would be to see if your current toolchain already has a built in solution that fits your needs. The examples in this category are somewhat artificially limited, as there are many tech stacks which account for having dynamic client-side behavior with server-side rendering. However, there's only a few cases where that hybrid behavior isn't at the core of how that technology functions. For example, it would be a rare case for someone to use NextJS without the explicit understanding that you naturally have access to client-side React when you need it. But there are some frameworks for which this kind of feature isn't on the front-page of the documentation.
```

Comment:
Consider adding a bit more context at the start of the article similar to this second sentence here.

"The examples outlined here are not meant to be exhaustive. A recent project with a stack similar to the ones outlined below had me researching these options as well as a few related options to fulfill our requirements. Hopefully my research and experimentation might aid in your own."

Reply:
Makes sense, and I see how it would benefit the spirit of the article. Went with...

> These scenarios and my recommended solutions are not entirely exhaustive, but recent work forced me to consider these scenarios more directly. Hopefully my experience and research can be an aid to you and your work